command-line-interface.md

```diff
@@ -0,0 +1,234 @@
+        * *stdout:* The runtime MUST attach its stdout directly to the container process without writing to it.
+        * *stderr:* The runtime MUST attach its stderr to the container process, and MUST not write to it unless it exits with a non-zero code.
+* *Environment variables*
+    * *`LISTEN_FDS`:* The number of file descriptors passed.
+      For example, `LISTEN_FDS=2` would mean that the runtime MUST pass file descriptors 3 and 4 to the container process (in addition to the standard streams) to support [socket activation][systemd-listen-fds].
+* *Exit code:* Zero if the container was successfully created and non-zero on errors.
+
+Callers MAY block on this command's successful exit to trigger post-create activity.
+
+#### Console socket
+
+The [`AF_UNIX`][unix-socket] [`SOCK_SEQPACKET`][socket-types] used by [`--console-socket`](#create) handles request and response messages between a client and server.
+The server MUST send a single response for each client request.
+The [normal data][socket-queue] ([`msghdr.msg_iov*`][socket.h]) of all messages MUST be [UTF-8][] [JSON](glossary.md#json).
+
+##### Requests
+
+All requests MUST contain a **`type`** property whose value MUST one of the following strings:
+
+* `terminal`, if the request is passing a [pseudoterminal master][posix_openpt.3].
+    When `type` is `terminal`, the request MUST also contain the following properties:
+
+    * **`container`** (string, REQUIRED) The container ID, as set by [create](#create).
+
+    The message's [ancillary data][socket-queue] (`msg_control*`) MUST contain a single [`cmsghdr`][socket.h]) with:
+
+    * `cmsg_type` set to [`SOL_SOCKET`][socket.h],
+    * `cmsg_level` set to [`SCM_RIGHTS`][socket.h],
+    * `cmsg_len` set to `CMSG_LEN(sizeof(int))`, and
+    * `*CMSG_DATA(cmsg)` set to the pseudoterminal master file descriptor.
+
+##### Responses
+
+All responses MUST contain a **`type`** property whose value MUST one of the following strings:
+
+* `success`, if the request was successfully processed.
+* `error`, if the request was not successfully processed.
+
+In addition, responses MAY contain any of the following properties:
+
+* **`message`** (string, OPTIONAL) A short phrase describing the response.
+
+#### Example
+
+```
+# in a bundle directory with a process that echos "hello" and exits 42
+$ test -t 1 && echo 'stdout is a terminal'
+stdout is a terminal
+$ funC create hello-1
+$ echo $?
+0
+$ funC start hello-1
+hello
+$ echo $?
+0
+$ block-on-exit-and-collect-exit-code hello-1
+$ echo $?
+42
+$ funC delete hello-1
+$ echo $?
+0
+```
+
+`hello` shows up in the terminal after `start`, because `start` happens to be called from the same shell session (using the same terminal) as `create`, and the container process inherited its standard streams from `create`.
+If you call `start` from a shell using a separate terminal, the container output would still have appeared in the `create` terminal.
```

Comment:
Example changed in 0edbe8f → 0530bda to decouple `create`'s stdio from the calling terminal.  And with that in place, I've dropped the whole explanatory paragraph.

I'm still not clear on “**blatantly** unsafe”, are you saying that there are still kernel bugs that allow privilege escalation via pty standard streams?

I'm also not clear on your dup'ing thoughts.  You *already* have a way to explicitly setup all the FDs: set `process.terminal` depending on whether you want a fresh pseudoterminal slave or inherited FDs on the standard streams, and then use `LISTEN_FDS` to list of the number of additional FDs to pass through.  And you can always use [`n<&-`][1] (if you're invoking `create` from a POSIX shell), [`O_CLOEXEC`][2], etc. to close off any FDs in that range that you don't want passed through.  And if you want to dup `/dev/null` over a file descriptor instead of just closing it (probably a good idea for the standard streams if you don't have more detail on the container process), you're already able to do that to without needing an additional option to set it up for you.

[1]: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_05
[2]: http://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html